Added ILU preconditioner

thoasm opened the pull request:
This PR adds an ILU preconditioner, which allows the customization of the Upper and the Lower solver separately. As soon as an `UpperTrs` exists, an alias can be created for the plain `Ilu` class with these two solvers as the default ones.

__TODO__:
- [X] Add test if `with_config` is present on solver before generating it
- [X] Find out how to do the reverse apply correctly
- [X] Add proper description for ILU
- [x] Merge #336 
- [x] Merge #341
- [x] Rebase
- [x] Replace default solvers in `Ilu` with `LowerTrs` and `UpperTrs` (LowerTrs works, tested with code from #336)

pratikvn commented:
If you do want to test it out. The UpperTrs class is here: [upper-trs-solver](https://github.com/ginkgo-project/ginkgo/tree/upper-trs-solver) and the cuda-kernels are here: [upper-trs-cuda-kernels](https://github.com/ginkgo-project/ginkgo/tree/upper-trs-cuda-kernels)

pratikvn commented:
Also just want to add that you should probably add creating a `factorization` algorithm to the `dev_tools/create_new_algorithm.sh` script.

thoasm commented:
@pratikvn Good idea, I added that to the TODOs

thoasm commented:
This PR should finally be ready for review.
However, PR #336 and #341 need to be merged before this one, so they are higher in priority.

pratikvn commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,377 @@
+                               .with_max_iters(default_max_iters)
+                               .on(exec),
+                           gko::stop::ResidualNormReduction<>::build()
+                               .with_reduction_factor(default_reduce_precision)
```
This should be `default_reduce_residual`. `default_reduce_precision` would mean something completely different.

pratikvn commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,377 @@
+        cache_struct() = default;
+        ~cache_struct() = default;
+        cache_struct(const cache_struct &other) {}
+        cache_struct &operator=(const cache_struct &) { return *this; }
```
I think the move construct is missing. Sonarqube will probably complain that you are not following rule of 5/0

thoasm replied:
Thanks, fixed it.

pratikvn commented on reference/test/preconditioner/ilu.cpp:
```diff
@@ -0,0 +1,317 @@
+        default_ilu_prec_type::build().on(exec)->generate(gko::share(par_ilu));
+    preconditioner->apply(b.get(), x.get());
+
+    // Since it is a preconditioner, the default solve is not accurate
```
In this case, it is exact :)

thoasm replied:
That is correct, I will change it after I rebase it. I needed a lower precision, so the three iterations and the rel. res. of 1e-4 would pass.

thoasm commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,377 @@
+          l_factor_{std::move(lu_args.l_factor)},
+          u_factor_{std::move(lu_args.u_factor)}
+    {
+        GKO_ASSERT_EQUAL_ROWS(l_factor_, u_factor_);
```
Good point, did that now.

thoasm commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,377 @@
+protected:
+    /**
+     * Manages the `generate` arguments for the parent class to allow multiple
+     * versions to initialize both L and U. Three constructors are provided:
```
My bad, I originally had 3, but I was not allowed to have a separate ParIlu one due to being ambiguous with Composition. Fixed it now.

thoasm commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,377 @@
+        GKO_ASSERT_EQUAL_ROWS(l_factor_, u_factor_);
+        GKO_ASSERT_EQUAL_COLS(l_factor_, u_factor_);
+
+        constexpr ValueType default_reduce_precision{1e-4};
```
Yes, I should have done that. Did it now, but not sure if the documentation is clear in the current state.

hartwiganzt commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,380 @@
+ * while providing the defaults solver::LowerTrs and solver::UpperTrs.
+ * For these solvers, a factory can be provided (with `with_l_solver_factory`
+ * and `with_u_solver_factory`) to have more control over their behavior.
+ * If these factories are not provided, the following criteria are set:
```
In particular, it is possible to use an iterative method for solving the triangular systems. The default parameters for an iterative triangular solver are:

hartwiganzt commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,380 @@
+ * If the solver does not support setting the criteria, none are provided.
+ *
+ * @note This class is not thread safe (even a const object is not) because it
+ *       uses an internal cache to accelerate multiple (sequential) applies
```
Please explain the consequences

thoasm replied:
I am actually not sure what all the possibilities are, so I just listed Segmentation fault and wrong result as potential consequences.

hartwiganzt commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,380 @@
+    /**
+     * @copydoc has_with_criteria
+     *
+     * @internal  The second template parameter (which uses SFINAE) must match
```
what is SFINAE

thoasm replied:
[SFINAE: Substitution Failure Is Not An Error](https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error), which means that if there is an invalid substitution of a template parameter, it does not count as a compiler error. It will act as if this function / structure / class does not exist, and continues the search for the best fit substitution (skipping every invalid substitution).

yhmtsai commented on reference/test/preconditioner/ilu.cpp:
```diff
@@ -0,0 +1,317 @@
+}
+
+
+TEST_F(Ilu, SolvesSingleRhsWithLU)
```
could it be `AppliesToVectorWithComposition`?
other tests for apply use this sentence.

thoasm replied:
Yes I could rename it from `Solves` to `Applies`, but I thought I go with the more specific description here.

The `ToVectorWithComposition` would not be accurate enough because I want to specify that it is a single vector, so a single right hand side. Also, in this test, it is with two separate matrices (L and U), the Composition would be the test above.

We could standardize it, but I don't think that there would be any real benefit from it. Currently, these names only help the developer to see which part does not work, so a descriptive name is very important. But apart from that, we don't have any standards / requirements for the test name as far as I know.

yhmtsai replied:
The name is good if we do not have the standard of this.

thoasm commented:
This can now be reviewed.
I added all the comments and added more tests in the hope of getting the coverage up.
On a note: I removed adding factorization to the script from the TODO, since it does not really fit here. I will create another PR to do that.

pratikvn commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,407 @@
+    /**
+     * Manages a vector as a cache, so there is no need to allocate one every
+     * time an intermediate vector is required.
+     * Copying an instance will only yield an empty object since copying the
```
if you dont want it to be copied, you could delete the copy and move constructors.

thoasm replied:
No, I can't, every PolymorphicObject needs to be copyable (see definition for [`copy_from`](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/polymorphic_object.hpp#L157)). In the end, this is called in [EnablePolymorphicAssignment](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/polymorphic_object.hpp#L616).
If it is not, it simply does not compile.

pratikvn replied:
Hm, you are right.

thoasm commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,407 @@
+        {
+            auto comp_cast =
+                as<const Composition<value_type>>(composition.get());
+            if (comp_cast->get_operators().size() != 2) {
```
What check do you mean? `as` already throws if the `dynamic_cast` was not successful.
See definition of [`as`](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/utils.hpp#L285).

pratikvn replied:
Ah, I didnt see that.

thoasm commented:
After the rebase, which contained a fix for the uploading artifact problem, all tests pass now.
This can definitively be reviewed now.

thoasm commented on core/test/preconditioner/ilu.cpp:
```diff
@@ -0,0 +1,98 @@
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/preconditioner/ilu.hpp>
+
```
What do you mean with `stlib` header? If you mean `memory`, then you are correct. I added that just now.

pratikvn replied:
Yes

pratikvn commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,407 @@
+ * solver are:
+ * - reduction factor = 1e-4
+ * - max iteration = <number of rows of the matrix given to the solver>
+ * Solvers without such criteria can also be used, in which case none are set.
```
I think you can add a mention that by default the solvers are direct triangular solvers. You have mentioned it in the `tparam` description, but I think it would be good to have it in the description as well.

thoasm replied:
Good point, I just added to line 65 (`while providing the defaults solver::LowerTrs and solver::UpperTrs`): `which are direct triangular solvers.`

pratikvn commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,407 @@
+        /**
+         * Returns the size that the solver using L and U would return
+         *
+         * @param inverse_apply  determines if the solver solves for U first
```
Why do you use two separate names ( `inverse_apply` and `ReverseApply`). It would be better to change this to `reverse_apply` as that is more suited, I think.

thoasm replied:
You are correct, it would. `inverse_apply` was my first name for it, then, I decided to go with `ReverseApply` for the template parameter, and forgot to change every mention of it.

tcojean commented on core/test/preconditioner/ilu.cpp:
```diff
@@ -0,0 +1,101 @@
+protected:
+    using value_type = double;
+    using Mtx = gko::matrix::Dense<value_type>;
+    using default_ilu = gko::preconditioner::Ilu<>;
```
I don't think you're using the three on top of this.

tcojean commented on core/test/preconditioner/ilu.cpp:
```diff
@@ -0,0 +1,101 @@
+    using u_solver_type = gko::solver::Bicgstab<value_type>;
+    using ilu_prec_type =
+        gko::preconditioner::Ilu<l_solver_type, u_solver_type, false>;
+    static constexpr int size_matrix{10};
```
Nor this one.

thoasm replied:
That is true, it seems I forgot to remove them when moving parts back to the reference test.
Thanks!

tcojean commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,410 @@
+
+
+/**
+ * The incomplete LU (ILU) preconditioner solves the equation LUx = b for a
```
minor: s/incomplete/Incomplete/

For math equations, you should check the documentation output but I think you would need to surround with `$`, e.g. for LUx = b

tcojean commented on include/ginkgo/core/preconditioner/ilu.hpp:
```diff
@@ -0,0 +1,410 @@
+         *
+         * @returns the size that the solver using L and U would return
+         */
+        dim<2> get_solver_size(bool reverse_apply = false) const
```
Should this be a parameter? Since you already have the reverse apply templated parameter?

thoasm replied:
You are right, the parameter is unnecessary. I removed it in the latest version.
Thanks!
